pkg/controllers/applicationfailover/rb_application_failover_controller.go

```diff
@@ -0,0 +1,249 @@
+		}
+		c.workloadUnhealthyMap.delete(resource)
+		return controllerruntime.Result{}, nil
+	}
+
+	retryDuration, err := c.syncBinding(binding)
+	if err != nil {
+		return controllerruntime.Result{Requeue: true}, err
+	}
+	if retryDuration > 0 {
+		klog.V(4).Infof("Retry to check health status of the workload after %v minutes.", retryDuration.Minutes())
+		return controllerruntime.Result{RequeueAfter: retryDuration}, nil
+	}
+	return controllerruntime.Result{}, nil
+}
+
+func (c *RBApplicationFailoverController) detectFailure(clusters []string, tolerationSeconds *int32, resource keys.ClusterWideKey) (int32, []string) {
+	var needEvictClusters []string
+	var duration int32
+
+	for _, cluster := range clusters {
+		if !c.workloadUnhealthyMap.hasWorkloadBeenUnhealthy(resource, cluster) {
+			c.workloadUnhealthyMap.setTimeStamp(resource, cluster)
+			duration = *tolerationSeconds
+		} else {
+			// When the workload in a cluster is in an unhealthy state for more than the tolerance time,
+			// and the cluster has not been in the GracefulEvictionTasks,
+			// the cluster will be added to the list that needs to be evicted.
+			unHealthyTimeStamp := c.workloadUnhealthyMap.getTimeStamp(resource, cluster)
+			if metav1.Now().After(unHealthyTimeStamp.Add(time.Duration(*tolerationSeconds) * time.Second)) {
+				needEvictClusters = append(needEvictClusters, cluster)
+			}
+		}
+	}
+
+	return duration, needEvictClusters
+}
+
+func (c *RBApplicationFailoverController) syncBinding(binding *workv1alpha2.ResourceBinding) (time.Duration, error) {
+	tolerationSeconds := binding.Spec.Failover.Application.DecisionConditions.TolerationSeconds
+	resource, err := helper.ConstructClusterWideKey(binding.Spec.Resource)
+	if err != nil {
+		klog.Errorf("failed to get key from binding(%s)'s resource", binding.Name)
+		return 0, err
+	}
+
+	allClusters := make(map[string]struct{})
+	for _, cluster := range binding.Spec.Clusters {
+		allClusters[cluster.Name] = struct{}{}
+	}
+
+	unhealthyClusters := filterIrrelevantClusters(binding.Status.AggregatedStatus, binding.Spec)
+	duration, needEvictClusters := c.detectFailure(unhealthyClusters, tolerationSeconds, resource)
+
+	err = c.evictBinding(binding, needEvictClusters)
+	if err != nil {
+		klog.Errorf("Failed to evict binding(%s/%s), err: %v.", binding.Namespace, binding.Name, err)
+		return 0, err
+	}
+
+	if len(needEvictClusters) != 0 {
+		if err = c.updateBinding(binding, allClusters, needEvictClusters); err != nil {
+			return 0, err
+		}
+	}
+
+	// Cleanup clusters that have been evicted or removed in the workloadUnhealthyMap
+	c.workloadUnhealthyMap.deleteIrrelevantClusters(resource, allClusters)
+
+	return time.Duration(duration) * time.Second, nil
+}
+
+func (c *RBApplicationFailoverController) evictBinding(binding *workv1alpha2.ResourceBinding, clusters []string) error {
+	for _, cluster := range clusters {
+		if binding.Spec.Failover.Application.PurgeMode == policyv1alpha1.Graciously {
```

Comment:
Can we use `switch` statement ?

Reply:
For the statement of three conditional branches, I think it's ok to use if-else here.

Reply:
Yes, switch-case would be more clear.